api/agent.go

```diff
@@ -60,12 +60,13 @@ type MembersOpts struct {
 
 // AgentServiceRegistration is used to register a new service
 type AgentServiceRegistration struct {
-	ID                string   `json:",omitempty"`
-	Name              string   `json:",omitempty"`
-	Tags              []string `json:",omitempty"`
-	Port              int      `json:",omitempty"`
-	Address           string   `json:",omitempty"`
-	EnableTagOverride bool     `json:",omitempty"`
+	ID                string            `json:",omitempty"`
+	Name              string            `json:",omitempty"`
+	Tags              []string          `json:",omitempty"`
+	Port              int               `json:",omitempty"`
+	Address           string            `json:",omitempty"`
+	EnableTagOverride bool              `json:",omitempty"`
+	ServiceMeta       map[string]string `json:",omitempty"`
 	Check             *AgentServiceCheck
 	Checks            AgentServiceChecks
```

Comment:
In order to prevent abusing this feature to store arbitarily long metadata values, add some validation. 

There is existing logic for node metadata here https://github.com/hashicorp/consul/blob/master/agent/structs/structs.go#L306. To keep things consistent we can use the same rules for service metadata as well. You should validate it in the agent and catalog registration endpoints. 

Its fine to allow the reserved consul prefix for service metadata since we don't rely on this for any internal uses.